### PR TITLE
feat(global): support vite inline import

### LINKED ISSUE #2533: the default export and sideEffects of `virtual:uno.css?inline`
### UnoCSS version

0.51.5

### Describe the bug

the module `virtual:uno.css?inline` still has sideEffects

image

---

when vite build, the default export of module `virtual:uno.css?inline` is `''`

image

image

---

when vite serve, there is not hmr for `virtual:uno.css?inline`

It should behave the same way as vite inline css

### Reproduction

<https://stackblitz.com/edit/vitejs-vite-hcufny?file=src%2Fmain.tsx&terminal=dev>

or

[test-shadow-dom/vite.config.ts](https://github.com/lisonge/vite-plugin-monkey/blob/614fd5d8c87a1e8112ebc6d130cd35521359d2bc/playground/test-shadow-dom/vite.config.ts)

### System Info

_No response_

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/unocss/unocss/blob/main/CONTRIBUTING.md).
- [X] Check that there isn't [already an issue](https://github.com/unocss/unocss/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/unocss/unocss/discussions) or join our [Discord Chat Server](https://chat.antfu.me).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

### PR DESCRIPTION
**Pull request by lisonge**
fix #2533

because of https://github.com/vitejs/vite/issues/12912, the HMR of the virtual module is unavailable by default at the moment

Before it is fixed,  it is also working by `/__uno.css?inline`

```ts
import unocssText from 'virtual:uno.css?inline';

const unocssStyle = document.createElement('style');
unocssStyle.dataset.source = 'unocss'
unocssStyle.textContent = unocssText;
export default unocssStyle;

if (import.meta.hot) {
    import.meta.hot.accept('/__uno.css?inline', (newModule) => {
        const text2 = newModule?.default as string;
        unocssStyle.textContent = text2 ?? ``;
    });
}
```

### CODE REVIEW

**Comment by enpitsuLin** [comment]
#2418 may relate to this, maybe `?inline` query and `?raw` should be the same behavior?

File: packages/vite/src/modes/global/build.ts [hunk 1]
```diff
@@ -100,7 +101,7 @@ export function GlobalModeBuildPlugin(ctx: UnocssPluginContext<VitePluginConfig>
         const layer = resolveLayer(getPath(id))
         if (layer) {
           vfsLayers.add(layer)
-          return ` ${getLayerPlaceholder(layer)}`
```
Comment (lisonge) [inline packages/vite/src/modes/global/build.ts]: @antfu hello, Can you explain to me why revert this line ?

[the remark this answers is not in the document]
Comment (antfu) [reply packages/vite/src/modes/global/build.ts]: I was about to ask why a space it added - I think adding a space is not the correct solution. Instead we should fix it in replacing.

File: packages/vite/src/modes/global/dev.ts [hunk 2]
```diff
@@ -197,13 +216,16 @@ try {
 if (!import.meta.url.includes('?'))
   await new Promise(resolve => setTimeout(resolve, 100))`
 
+          // Reload full page when using `?inline` - any better solution?
+          if (isInlineCssMod)
+            hmr += '\nimport.meta.hot.accept(() => location.reload())\n'
```
[the remark this answers is not in the document]
Comment (antfu) [reply packages/vite/src/modes/global/dev.ts:221]: `reloadModule` is a very new API and does not available in all Vite versions. + we should avoid making side-effects inside `.map` whenever possible. If it's necessary, we should 

1. Check if `reloadModule` exists on server, and fallback to custom logic when using with lower version
2. Move the calls outside of `.map`, using `.forEach` or others.


---

I actually found either solution works for HMR. Did you manually write the HMR handling?

**Comment by antfu** [comment]
I think the solution is incomplete, and I need to find another time to do a full revisit. Marked as a draft for now. Feel free to push any changes you think making sense.

**Comment by lisonge** [comment]
A simple way to customize unocss **global** style custom side effect

it work well with hmr and shadow-dom

[test-example](https://github.com/lisonge/vite-plugin-monkey/tree/47ac609752120b62cf8a8d5f22b9f7d184c4c59e/playground/test-unocss)

Although there may be defects, it meets my needs

```ts
import { build } from '@unocss/cli';
import fs from 'node:fs/promises';
import unocss from 'unocss/vite';
import type { ConfigEnv, Plugin } from 'vite';

const iframeTemplate = /* html */ `
<script type="module">
// base must be '/'
import '/@vite/client';
import '/__uno.css';

const unoStyle = document.querySelector(
  'head style[data-vite-dev-id="/__uno.css"]',
);
const update = ()=>{
  parent.postMessage(
    {
      type: '__unocss_update',
      textContent: unoStyle?.textContent,
    },
    '*',
  );
}
if (window != parent) {
  update();
  new MutationObserver(update).observe(unoStyle, {
    childList: true,
    attributes: true,
  });
}
</script>
`.trim();

const moduleTemplate = /* js */ `
const style = document.createElement('style');
style.textContent = __UNOCSS_TEXT__;
export default style;
if (import.meta.env.DEV) {
  const origin = new URL(import.meta.url).origin;
  window.addEventListener('message', (event) => {
    const data = event.data || {};
    if (data.type == '__unocss_update') {
      style.textContent = data.textContent;
    }
  });
  const iframe = document.createElement('iframe');
  iframe.src = origin + '/__unocss_iframe';
  iframe.style.display = 'none';
  document.head.append(iframe);
}
`.trim();

export const unocssStyle = async (env: ConfigEnv): Promise<Plugin[]> => {
  const dev = env.command == 'serve';
  const unocss_text = await (async () => {
    if (!dev) {
      // or use createGenerator({},{}).generate(code)
      await build({
        cwd: process.cwd(),
        config: './unocss.config.ts',
        patterns: ['./src/**/*.{js,mjs,ts,jsx,tsx,vue,svelte}'],
        stdout: false,
        minify: true,
        outFile: './uno-prebuilt.css',
      });
      const text = (await fs.readFile('./uno-prebuilt.css', 'utf-8')).trim();
      await fs.unlink('./uno-prebuilt.css');
      return text;
    } else {
      return '';
    }
  })();

  return [
    {
      name: 'unocss:style',
      resolveId(source) {
        if (source == `unocss?style`) return `__unocss_style`;
      },
      async load(id) {
        if (id == `__unocss_style`) {
          return moduleTemplate.replace(
            '__UNOCSS_TEXT__',
            JSON.stringify(unocss_text),
          );
        }
      },
      configureServer(server) {
        server.middlewares.use(async (req, res, next) => {
          if (req.url?.startsWith(`/__unocss_iframe`)) {
            res.setHeader('content-type', 'text/html');
            res.end(iframeTemplate);
            return;
          }
          next();
        });
      },
    },
    ...(dev ? unocss() : []),
  ];
};
```

```tsx
// main.tsx
import { render, Portal } from 'solid-js/web';
import style from 'unocss?style';
import App from './App';

render(
  () => (
    <Portal useShadow>
      <App />
      {style}
    </Portal>
  ),
  (() => {
    const div = document.createElement('div');
    document.body.append(div);
    return div;
  })(),
);

```

**Comment by lisonge** [comment]
https://github.com/lisonge/unocss-inline

**Pull request closed without merging**